ocis: add sequential upgrade path through 8.1

Ramblurr opened the pull request:
Edit 2026-07:

This PR has been languishing too long, it's over a year old.

I rebuilt this PR on current master and expanded it to cover the complete upstream production upgrade sequence:

 `5.0 → 7.0 → 7.1 → 7.2 → 7.3 → 8.0 → 8.1`

Existing systems will stay on `ocis_5-bin`, while new NixOS 26.11 systems default to `ocis_81-bin`.
---

Original PR:

This is a follow up to https://github.com/NixOS/nixpkgs/pull/394598

TL;DR: (full explanation is in the "Maintainer information" in the nixos documentation)

Upgrading ocis requires stepping through every stable major *and* minor version. You cannot skip one.

This PR adds ocis 7.0 and ocis 7.1. ocis 6.0 was skipped as it is not a stable version.

The nixos module is updated to choose the appropriate version based on the
system.stateVersion, and now provides some tools to help the user update their
instance.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

## pinging fellow maintainers

@bhankas @danth

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ramblurr commented:
Fixed merge conflicts

Ramblurr commented:
Fixed merge conflicts, again.

Ramblurr commented:
Fixed merge conflicts.

Ramblurr commented:
Fixed new fmt requirements.

Ramblurr committed:
nixos/ocis: support sequential upgrades

Preserve the 5.0 default for existing systems and provide every production
upgrade step through 8.1 for new NixOS 26.11 installations.

Remove bhankas from the module maintainers at their request in the PR
discussion.

https://doc.owncloud.com/ocis/next/migration/upgrading-ocis.html

dotlambda reviewed (commented):
It seems to me like the different derivativions share most of their code. Can we make them call a generic builder instead to deduplicate that?

Ramblurr committed:
ocis-bin: init 7.0.1 through 8.1.0

https://github.com/owncloud/ocis/releases/tag/v7.0.1
https://github.com/owncloud/ocis/releases/tag/v7.1.4
https://github.com/owncloud/ocis/releases/tag/v7.2.0
https://github.com/owncloud/ocis/releases/tag/v7.3.2
https://github.com/owncloud/ocis/releases/tag/v8.0.6
https://github.com/owncloud/ocis/releases/tag/v8.1.0

Ramblurr commented:
I have decided not to pursue oCIS in nixpkgs any further and am retracting this PR.
